aea/agent_loop.py

```diff
@@ -0,0 +1,406 @@
+import asyncio
+import datetime
+import logging
+import time
+from abc import ABC, abstractmethod
+from asyncio.events import AbstractEventLoop, TimerHandle
+from asyncio.futures import Future
+from asyncio.tasks import ALL_COMPLETED, FIRST_COMPLETED, Task
+from typing import Any, Callable, Dict, List, Optional, Set, Tuple, Union, cast
+
+from aea.exceptions import AEAException
+from aea.mail.base import InBox
+from aea.skills.base import Behaviour
+
+
+try:
+    from asyncio import create_task
+except ImportError:
+    # for python3.6!
+    from asyncio import ensure_future as create_task  # type: ignore
+
+
+logger = logging.getLogger(__file__)
+
+if False:  # MYPY compatible for types definitions
+    from aea.aea import AEA
+    from aea.agent import Agent
+
+
+def ensure_list(value: Any) -> List:
+    """Return [value] or value if value is a list."""
+    if not isinstance(value, list):
+        value = [value]
+    return value
+
+
+class AsyncState:
+    """Awaitable state."""
+
+    def __init__(self, initial_state: Any = None, loop: AbstractEventLoop = None):
+        """Init async state.
+
+        :param initial_state: state to set on start.
+        :param loop: optional asyncio event loop.
+        """
+        self._state = initial_state
+        self._watchers: Set[Future] = set()
+        self._loop = loop or asyncio.get_event_loop()
+
+    @property
+    def state(self) -> Any:
+        """Return current state."""
+        return self._state
+
+    @state.setter
+    def state(self, state: Any) -> None:
+        """Set state."""
+        if self._state == state:
+            return
+        self._state_changed(state)
+        self._state = state
+
+    def _state_changed(self, state: Any) -> None:
+        """Fulfill watchers for state."""
+        for watcher in list(self._watchers):
+            if state in watcher._states:  # type: ignore
+                self._loop.call_soon_threadsafe(
+                    watcher.set_result, (self._state, state)
+                )
+
+    async def wait(self, state_or_states: Union[Any, List[Any]]) -> Tuple[Any, Any]:
+        """Wait state to be set.
+
+        :params state_or_states: state or list of states.
+        :return: tuple of previous state and new state.
+        """
+        states = ensure_list(state_or_states)
+
+        if self._state in states:
+            return (None, self._state)
+
+        watcher: Future = Future()
+        watcher._states = states  # type: ignore
+        self._watchers.add(watcher)
+        try:
+            return await watcher
+        finally:
+            self._watchers.remove(watcher)
+
+
+class PeriodicCaller:
+    """Schedule a periodic call of callable using event loop."""
+
+    def __init__(
+        self,
+        callback: Callable,
+        period: float,
+        start_at: Optional[datetime.datetime] = None,
+        exception_callback: Optional[Callable] = None,
```

Comment:
maybe specify which arguments the callable expects?

From the code below I can see it expects:
- the callback (hence, of type `Callable`
- the exception (type `Exception`)

so the full type should be:
```
exception_callback: Optional[Callable[[Callable], Exception]] = None,
```

do you agree?

Reply:
yes, it's a nice point
updated.